### PR TITLE
wallet-ext: improve gas estimations for nft and coin transfers 

### PR DESCRIPTION
**Pull request by pchrysochoidis**
Draft to work on splitting this PR to smaller ones and improve gas estimation api

* [wallet-ext: fix crypto mock in test setup](https://github.com/MystenLabs/sui/pull/6156/commits/25666d51e5f65a4e0def6f0203c5f8c1be3cea73) 
  * it breaks the tests since now crypto seems to be defined
* [ts-sdk: adds method to estimate gas cost and suggested budget to signer](https://github.com/MystenLabs/sui/pull/6156/commits/1ab09c3278920cd9115a297dd4e40be4fdbc5775) 
 * also fixes some edge cases with pay txs (take into account that only the first coin is used for gas)
* [wallet-ext: improve gas estimations for nft transfer](https://github.com/MystenLabs/sui/pull/6156/commits/6e457dcf00cb6ee4464fe336b381fc43594638cc) 
  * fixes edge cases and uses new method from signer
  * use Alert component for errors
* [wallet-ext: improve token transfer gas budget estimations](https://github.com/MystenLabs/sui/pull/6156/commits/b8174d965d00d95e25cbda2e2f6fe3a1de6714d0) 
  * uses new method from signer to estimate gasCost and suggested gasBudget
  * updates token validation to include more cases
  * uses Alert for errors in send tokens form
  
Screenshot 2022-11-18 at 22 33 42
Screenshot 2022-11-18 at 22 34 36
Screenshot 2022-11-18 at 22 34 48

closes APPS-149

### CODE REVIEW

File: sdk/typescript/src/signers/signer-with-provider.ts
Comment (gegaowp) [inline sdk/typescript/src/signers/signer-with-provider.ts:418]: I find this forError is confusing, can we derive it from if maxGasCoinBalance is null or not?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (gegaowp) [inline sdk/typescript/src/utils/gas-estimation.ts:67]: same here _forError seems not necessary?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (gegaowp) [inline sdk/typescript/src/utils/gas-estimation.ts:83]: is gasBudgetGuess < MAX_GAS_BUDGET always true? if so, looks like we do not need MAX_GAS_BUDGET here?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (gegaowp) [inline sdk/typescript/src/utils/gas-estimation.ts:72]: why do we want to take maxGasCoinBalance for gas estimation? imo
- estimation is only about txn to be run, if estimation itself failed, we want to fall back to use some constants or numbers in tables as the estimated gas 
- if maxGasCoinBalance < estimated gas (either from dry run or from constants), we want to trigger some preparation steps to make a gas coin for the txn.

**Review by gegaowp — COMMENTED** [review]
Great move! I left several comments on the SDK side, will leave apps team for the apps/ dir.

File: sdk/typescript/src/signers/signer-with-provider.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/signers/signer-with-provider.ts:418]: Yeah you are right, I will change it to that. I introduced this value to notify `getGasBudgetGuess` (and make it more intentional) that we don't have dry run or something didn't work (with dry-run) so make a guess based on that (do not optimise the guess based on minimum available coins and max IndividualCoinBalance).

Probably removing it and deriving it from `maxGasCoinBalance` should be ok.

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:67]: I will remove it, can think of something better when and if we need it

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:72]: We are taking `maxGasCoinBalance` into account because that's the minimum budget (not always - in case we send sui) the user can pay. And if we get insufficient funds then we know we need to merge coins. Using as budget only the local constants does't help to improve the estimate.

For example if we set the budget to 1000 and the `maxGasCoinBalance` is  400 the dry run will fail. But the actual cost/budget could be 300. Using `maxGasCoinBalance` helps to improve estimate in this case.

I am not sure if possible, but ideally if we could ignore budget/cost restrictions for dry run would make it easier to use it. To return something like "if you had the gas this would cost x"

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:83]: For the current cases yes - I would leave it here as a 'safety' check in case we miss it anywhere else

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts]: In the end removed fns for keys that didn't actually had to be fns

**Comment by pchrysochoidis** [comment]
Thanks for the review @gegaowp, I think I addressed the comments, let me know if there is anything else!

**Review by gegaowp — APPROVED** [review]
The SDK part looks great to me, AC to unblock! Note that I did not really review the apps/ closely, would be great if we have a second pair of eyes on that.

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (Jordan-Mysten) [inline sdk/typescript/src/utils/gas-estimation.ts:72]: Why can't this use dry-run?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (Jordan-Mysten) [inline sdk/typescript/src/utils/gas-estimation.ts:19]: Can we document both of these magic numbers?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (Jordan-Mysten) [inline sdk/typescript/src/utils/gas-estimation.ts:21]: I assume this needs to be kept in sync with some backend value, can we document that?

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (Jordan-Mysten) [inline sdk/typescript/src/utils/gas-estimation.ts:29]: I'm confused why this can't be done with dry run.

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (Jordan-Mysten) [reply sdk/typescript/src/utils/gas-estimation.ts:72]: Why not just set it to something very high when dry-running?

@gegaowp I'm curious if we can still succeed for dry run even when gas budget is exceeded.

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:19]: I figured them out by running txs with too small and too big gas budgets. What should we document? Their meaning? or how we calculated them?

(They don't really affect the calculations for budget only for some edge cases)

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:21]: I guess, moved this from Coin so not sure why is this number, I will add a comment

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:29]: Same as the above, and also in case of dry-run fails we use this estimation

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:72]: > Why not just set it to something very high when dry-running

Then we have the problem that tx can fail because of insufficient funds, but still we need to handle the case dry-run fails. We could simply set it to the value of the coin with max balance (-amount in case we send Sui, etc) which is what we are doing except we start from a max gas budget that makes sense for the tx type

Comment (Jordan-Mysten) [reply sdk/typescript/src/utils/gas-estimation.ts:72]: This makes more sense to me. I really do think we need to fix this at the dry run level, and not paper over it like this. I definitely don’t think these should be in the SDK, as ideally we can just fix dry run. As it stands I think it’s way too complicated for partners to actually estimate gas, and we should fix that. I read the comments on the other thread too, seems like something that is solvable. I think we should even if it’s hard. 

I get the case where the dry run fails. I think in those cases we either use a default max value, and warn the user their transaction may fail, and that it may use a lot of gas. IMO we need to guess it for those cases, we just need to tell the user. Ideally dry run failing is actually our signal that the tx itself will fail and we’ll actually gate the default transaction UI on that.

Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:72]: I agree changing dry-run to ignore gas budget would help, the only thing that will change though would be the removal of the `optimizeForDryRun` and probably the `maxGasCoinBalance` of this function. We still need for some cases the guess constant to be a function, because gas depends on number of coins etc.

This function is more meant as an internal to the sdk utility - users should use the [getGasCostEstimationAndSuggestedBudget](https://github.com/MystenLabs/sui/pull/6156/files/5bd6e33ba5f0556ca90cb5ef3bf1feffbf56be30#diff-90a054a3b632795af85582441a61df5784805807e652e8167d61978a198a17faR385) of the signer where it returns the estimation (either from dry-run or from guess).

We can easily remove the extra functionality, to optimise for dry-run when dry run changes and that should not affect much the end user.

I think [getGasCostEstimationAndSuggestedBudget](https://github.com/MystenLabs/sui/pull/6156/files/5bd6e33ba5f0556ca90cb5ef3bf1feffbf56be30#diff-90a054a3b632795af85582441a61df5784805807e652e8167d61978a198a17faR385) works more or less how you describe above, maybe it need some changes to it's api.

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:27]: done

File: sdk/typescript/src/utils/gas-estimation.ts
Comment (pchrysochoidis) [reply sdk/typescript/src/utils/gas-estimation.ts:33]: done

**Comment by pchrysochoidis** [comment]
all changes except those that are related to gas estimation using dryRun were split to other PRs https://github.com/MystenLabs/sui/pull/6695, https://github.com/MystenLabs/sui/pull/6701, https://github.com/MystenLabs/sui/pull/6703, https://github.com/MystenLabs/sui/pull/6706. Gas estimation is blocked until gasBudget and gas cost can be ignored.

**Pull request closed without merging**
